### ISSUE TITLE
CRASH: Illegal Instruction

### ISSUE THREAD

**Issue by maroueneboubakri** [DynamoRIO/dynamorio#4319]
my.bin is an ARM  binary of size 5MB 
After w while all crash with Illegal Instruction. 
Same behavior is seen with all samples.  Either with a client which does nothing.
Version: DynamoRIO-ARM-Linux-EABIHF-8.0.18417

```
# gdb ./bin32/drrun

(gdb) r -c samples/build/bin/libdiv.so -- /my.bin
Starting program: /root/workspace/dynamorio/lib32/release/libdynamorio.so -c samples/build/bin/libdiv.so -- /my.bin

Program received signal SIGILL, Illegal instruction.
0x7f61df8c in syscall_0_or_1arg ()
(gdb) x/10i syscall_0_or_1arg
   0x7f61df88 <syscall_0_or_1arg>:      svc     0x00000000
=> 0x7f61df8c <syscall_0_or_1arg+4>:    pop     {r4, r5, r6, r7, r8, r9}
   0x7f61df90 <syscall_0_or_1arg+8>:    bx      lr
   0x7f61df94 <dr_fpu_exception_init>:  bx      lr
   0x7f61df98 <is_shebang>:     ldrb    r3, [r0, #2]
   0x7f61df9a <is_shebang+2>:   adds    r0, #2
   0x7f61df9c <is_shebang+4>:   cmp     r3, #9
   0x7f61df9e <is_shebang+6>:   it      ne
   0x7f61dfa0 <is_shebang+8>:   cmpne   r3, #32
   0x7f61dfa2 <is_shebang+10>:  bne.n   0x7f61dfb0 <is_shebang+24>
(gdb) quit
```
Thanks
Maro

**Comment by derekbruening** [comment]
DR tries very hard to be transparent.  You can see extensive documentation on that in the tutorial, at https://dynamorio.org/dynamorio_docs/transparency.html, and in the academic literature.  If there is a particular violation here I would suggest detailing exactly what the application is doing to cause the problem.  Once the details are known, then potential solutions can be formulated.

**Comment by maroueneboubakri** [comment]
Your support was appreciated thank you I close the ticket :)

**Comment by derekbruening** [comment]
I'm confused: why close it?  You've found a case where DR does not properly execute an application: it doesn't reproduce the native behavior, right?  It's a bug in DR?

**Comment by derekbruening** [comment]
We're happy to help formulate solutions to fix the problem, if you could provide details on what the application is doing to cause the problem.

**Comment by maroueneboubakri** [comment]
Actually I tried drmemory and worked fine, meaning that should consider more cases in my client. The sample clients were not helpful though. Should dig more with what Drmemory does and follow the pattern.

**Comment by derekbruening** [comment]
> Actually I tried drmemory and worked fine, meaning that should consider more cases in my client. 

?  But you said this app crashes under plain DR and all the sample clients.  You're saying this same app runs under drmemory?

**Comment by maroueneboubakri** [comment]
Correct, It fails with all samples but it succeeds to execute with drmemory. I'm still investigating the reason. 

The following dumb client works:

```
#include "dr_api.h"

static void
event_exit(void);


DR_EXPORT void
dr_client_main(client_id_t id, int argc, const char *argv[])
{
    dr_set_client_name("DynamoRIO Sample Client 'working'", "http://dynamorio.org/issues");
    dr_register_exit_event(event_exit);
}

static void
event_exit(void)
{
}

```


The following minimalist client fails with SIGILL,it seems that drmgr_init is doing something which my binary does not like.

```
#include "dr_api.h"
#include "drmgr.h"

static void
event_exit(void);


DR_EXPORT void
dr_client_main(client_id_t id, int argc, const char *argv[])
{
    dr_set_client_name("DynamoRIO Sample Client 'working'", "http://dynamorio.org/issues");
	drmgr_init();
    dr_register_exit_event(event_exit);
}

static void
event_exit(void)
{
	drmgr_exit();
}

```

**Comment by maroueneboubakri** [comment]
It seems that Drmemory is doing some magic to handle SIGILL here https://github.com/DynamoRIO/drmemory/blob/master/drmemory/pattern.c#L955
I will try to grab this code to my sample and see what it brings.

**Comment by derekbruening** [comment]
> It seems that Drmemory is doing some magic to handle SIGILL here https://github.com/DynamoRIO/drmemory/blob/master/drmemory/pattern.c#L955
> I will try to grab this code to my sample and see what it brings.

That's a tool-generated SIGILL, not an app-generated one.

**Comment by maroueneboubakri** [comment]
Well, by taking the SIGILL handler from drmemory everything was fine. Seems that switching between THUMB ISA and ARM32 is not handled naively by DR and should be handled by client developer right ?

**Comment by derekbruening** [comment]
> Seems that switching between THUMB ISA and ARM32 is not handled naively by DR and should be handled by client developer right ?

DR certainly does handle an application switching between Thumb and AArch32.  It has to, or else most 32-bit ARM binaries would not work.
